### PR TITLE
Branch sum dfs mock interview q

### PR DESCRIPTION
**Pull request by abethegr8**
Mock Interview Question, find the sum of each branch.

### CODE REVIEW

File: src/BranchSum.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if (current == null) {
+            return returnSum;
+        }
+        dfs(current, sum, returnSum);
+
+        return returnSum;
+
+    }
+
+    public static void dfs(Node current, int sum, List<Integer> returnSum){
```
Comment (jchaffraix) [inline src/BranchSum.java:69]: Prefer names that help the reader understand what the function is doing as it is not a generic depth first search function: it is very tied to the problem at hand.

Some ideas for better names: `branchSumRecursive`. `branchSumDepthFirstSearch`

Also this should be `private`: we do not want users outside of this class to use it. They should go through the public function `branchSum`.

File: src/BranchSum.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        Node(int val, Node left, Node right) {
+            this.val = val;
+            this.left = left;
+            this.right = right;
+        }
+    }
+
+    public static List<Integer> branchSum(Node root){
+        List<Integer> returnSum = new ArrayList<>();
+        int sum = 0;
```
Comment (jchaffraix) [inline src/BranchSum.java:58]: You should define variables when they are needed. Here are have a early return so this should be moved after it, just above l.63.

Also good to realize that this variable is actually not needed (except for readability).

File: src/BranchSum.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        Node current = root;
```
Comment (jchaffraix) [inline src/BranchSum.java:59]: `current` is not needed. If anything, it obscures the code as a reader needs to remember that `current == root` in this block.

File: src/GroupAnagrams.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
+/*Leetcode: https://leetcode.com/problems/group-anagrams/
```
Comment (jchaffraix) [inline src/GroupAnagrams.java:1]: This belongs to the other pull request (PR #4). Did you start this branch from the other branch by any chance?

If that's the case, you will want to rebase your branch on top of master, which should remove this:
```
git checkout BranchSumDFS_MockInterviewQ
git rebase master
```

I don't know how to do those steps on IntelliJ unfortunately. Try to fix it and we can look into it together next time if you're stuck.

File: src/BranchSum.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        Node() {}
+        Node(int val) { this.val = val; }
+        Node(int val, Node left, Node right) {
+            this.val = val;
+            this.left = left;
+            this.right = right;
+        }
+    }
+
+    public static List<Integer> branchSum(Node root){
```
Comment (jchaffraix) [inline src/BranchSum.java:56]: This is missing the big-O complexities.

**Review by jchaffraix — COMMENTED** [review]
Also the description is missing for this PR. I recommended filling it:

empty_description

File: src/BranchSum.java (hunk 1 above)
Comment (abethegr8) [reply src/BranchSum.java:69]: This is great advice. I have updated the helper function to be a private function and updated its name to help the reader understand the function.

File: src/BranchSum.java (hunk 2 above)
Comment (abethegr8) [reply src/BranchSum.java:58]: Got it, updated.

File: src/GroupAnagrams.java (hunk 4 above)
Comment (abethegr8) [reply src/GroupAnagrams.java]: We fixed this during our meeting together on 7/27.

File: src/BranchSum.java (hunk 5 above)
Comment (jchaffraix) [reply src/BranchSum.java:56]: FYI This is still missing, but I think we went over this together.

File: src/BranchSum.java (hunk 3 above)
Comment (jchaffraix) [reply src/BranchSum.java:58]: This still applies.

**Review by jchaffraix — APPROVED** [review]
I think that's good enough. I would prefer if we fixed the comment about `current`.

**Comment by abethegr8** [comment]
PR Approved, merging.

**Pull request merged**